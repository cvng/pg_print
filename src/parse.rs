use crate::fmt::Printer;
use crate::stmt::RawStmt;

impl Printer {
    #[cfg(feature = "unstable")]
    pub fn parse(&mut self, parse: &parser::Parse) {
        self.cbox(0);
        for stmt in &parse.stmts {
            self.stmt(&RawStmt { stmt: &stmt.stmt });
        }
        self.end();
    }

    #[cfg(not(feature = "unstable"))]
    pub fn parse(&mut self, parse: &pg_query::protobuf::ParseResult) {
        self.cbox(0);
        for stmt in &parse.stmts {
            if let Some(stmt) = &stmt.stmt {
                if let Some(node) = &stmt.node {
                    self.stmt(&RawStmt { stmt: node });
                }
            }
        }
        self.end();
    }
}
