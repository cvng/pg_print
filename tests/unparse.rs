// Adapted from https://github.com/lelit/pglast/blob/v5/tests/test_printers_prettification.py.

use insta::assert_snapshot;
use pg_query::parse;
use rstest::rstest;
use std::env;
use std::fs;
use std::path::PathBuf;

macro_rules! set_snapshot_suffix {
    ($($expr:expr),*) => {
        let mut settings = insta::Settings::clone_current();
        settings.set_snapshot_suffix(format!($($expr,)*));
        let _guard = settings.bind_to_scope();
    }
}

#[rstest]
fn unparse(
    #[files("tests/**/*.sql")]
    #[exclude("exclude")]
    path: PathBuf,
) {
    let path = path.strip_prefix(env::current_dir().unwrap()).unwrap();

    for (mut lineno, case) in fs::read_to_string(path)
        .unwrap()
        .split('\n')
        .filter(|line| !line.is_empty())
        .filter(|line| !line.starts_with("--"))
        .enumerate()
    {
        lineno += 1;

        set_snapshot_suffix!("{}", lineno);

        let deparsed = pg_query::deparse(&parse(case).unwrap().protobuf)
            .unwrap()
            .trim_end()
            .to_string();

        let unparsed = pg_deparser::unparse(&parse(case).unwrap().protobuf)
            .unwrap()
            .trim_end()
            .to_string();

        let reparsed = pg_query::deparse(&parse(&unparsed).unwrap().protobuf)
            .unwrap()
            .trim_end()
            .to_string();

        assert_eq!(deparsed, reparsed);

        assert_snapshot!(
            path.to_str().unwrap().replace(['/', '.'], "_"),
            unparsed,
            &format!("{}:{}", path.display(), lineno)
        );
    }
}
