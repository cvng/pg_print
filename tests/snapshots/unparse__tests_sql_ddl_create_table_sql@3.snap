---
source: tests/unparse.rs
expression: "tests/sql/ddl/create_table.sql:3"
---
create temporary table a (
    id serial
) on commit drop
